File: app/src/main/java/com/miaadrajabi/menuxmlparser/java/presentation/menu/FlowObserver.java
package com.miaadrajabi.menuxmlparser.java.presentation.menu;

import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.Observer;

public class FlowObserver<T> implements DefaultLifecycleObserver {
    private final Lifecycle lifecycle;
    private final androidx.lifecycle.LiveData<T> liveData;
    private final Observer<T> observer;

    public FlowObserver(Lifecycle lifecycle, androidx.lifecycle.LiveData<T> liveData, Observer<T> observer) {
        this.lifecycle = lifecycle;
        this.liveData = liveData;
        this.observer = observer;
    }

    @Override
    public void onStart(LifecycleOwner owner) {
        liveData.observe(owner, observer);
    }

    @Override
    public void onStop(LifecycleOwner owner) {
        liveData.removeObserver(observer);
    }
}

File: app/src/main/java/com/miaadrajabi/menuxmlparser/presentation/menu/FlowObserver.kt
package com.miaadrajabi.menuxmlparser.presentation.menu

import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class FlowObserver<T>(
    private val lifecycle: Lifecycle,
    private val flow: StateFlow<T>,
    private val block: (T) -> Unit
) : DefaultLifecycleObserver {
    private var job: Job? = null

    override fun onStart(owner: LifecycleOwner) {
        job = owner.lifecycleScope.launch {
            flow.collect { block(it) }
        }
    }

    override fun onStop(owner: LifecycleOwner) {
        job?.cancel()
    }
}


